File: Session_2/Bài 3.2 Quick Short/Quick_Short.jl
# Đây là chương trình Julia sử dụng thuật toán Quick Sort để sắp xếp mảng
function partitionFunc(arr, low::Int , high::Int)
    pivot = arr[high]  # Chọn phần tử cuối cùng làm pivot
    i = low - 1  # Khởi tạo chỉ số cho phần tử nhỏ hơn pivot
    for j in low:high-1  # Duyệt qua các phần tử trong mảng
        if arr[j] < pivot  # Nếu phần tử nhỏ hơn pivot
            i += 1  # Tăng chỉ số i
            arr[i], arr[j] = arr[j], arr[i]  # Hoán đổi phần tử
        end
    end
    arr[i+1], arr[high] = arr[high], arr[i+1]  # Đưa pivot vào vị trí đúng
    return i+1  # Trả về chỉ số của pivot
end

function quickSort!(arr, low::Int, high::Int)
    if low < high  # Nếu mảng có ít nhất 2 phần tử
        pi = partitionFunc(arr, low, high)  # Phân vùng mảng
        quickSort!(arr, low, pi-1)  # Sắp xếp phần mảng bên trái của pivot
        quickSort!(arr, pi+1, high)  # Sắp xếp phần mảng bên phải của pivot
    end
    return arr  # Trả về mảng đã sắp xếp
end

# Kiểm tra chương trình
arr = [5, 2, 9, 1, 5, 6]
quickSort!(arr, 1, length(arr))  # Đánh số từ 1, gọi quickSort
println("Kết quả Quick Sort: ", arr)  # Output: [1, 2, 5, 5, 6, 9]

File: Session_2/Bài 3.2 Quick Short/Quick_Short.cpp
//đây là chương trình sắp xếp mảng dựa trên thuật toán quick short
#include <iostream>
#include <vector>

using namespace std;

// Hàm Partition
int partitionFunc(vector<int> &arr, int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) // Duyệt qua mảng từ `low` đến `high - 1`
    {
        if (arr[j] < pivot)
        {
            i++;
            // Hoán đổi thủ công không dùng std::swap
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    // Hoán đổi pivot với phần tử đứng sau `i`
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;

    return i + 1; // Trả về vị trí của pivot
}

// Hàm Quick Sort
void quickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        // Tìm pivot sao cho mảng được chia thành 2 mảng con
        int pi = partitionFunc(arr, low, high);

        // Đệ quy sắp xếp 2 mảng con
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main()
{
    vector<int> arr = {5, 2, 9, 1, 6, 5};
    int n = arr.size();
    quickSort(arr, 0, n - 1);

    cout << "Sorted array: \n";
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}
